src/csharp/Grpc.Microbenchmarks/CompletionRegistryBenchmark.cs

```diff
@@ -0,0 +1,69 @@
+﻿#region Copyright notice and license
+
+// Copyright 2015 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Grpc.Core;
+using Grpc.Core.Internal;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Grpc.Microbenchmarks
+{
+    public class CompletionRegistryBenchmark
+    {
+        GrpcEnvironment environment;
+
+        public void Init()
+        {
+            environment = GrpcEnvironment.AddRef();
+        }
+
+        public void Cleanup()
+        {
+            GrpcEnvironment.ReleaseAsync().Wait();
+        }
+
+        public void Run(int threadCount, int iterations)
+        {
+            Console.WriteLine(string.Format("CompletionRegistryBenchmark: threads={0}, iterations={1}", threadCount, iterations));
+            var threadedBenchmark = new ThreadedBenchmark(threadCount, () => ThreadBody(iterations));
+            threadedBenchmark.Run();
+            // TODO: parametrize by number of pending completions
+        }
+
+        private void ThreadBody(int iterations)
+        {
+            var completionRegistry = new CompletionRegistry(environment);
+            var ctx = BatchContextSafeHandle.Create();
+            var completionDelegate = new OpCompletionDelegate((success) => {});
+  
+            var stopwatch = Stopwatch.StartNew();
+            for (int i = 0; i < iterations; i++)
```

Comment:
since it looks like the whole loop here will finish with a net zero for its modifying of external state, can we be sure that things aren't getting optimized away here?

If that's the case, perhaps e.g., `callback` can get called after being extracted, `completionDelegate` can increment some counter, and then that counter can be checked after the loop is finished?

Reply:
I think you're right that there's a theoretical change of this getting optimized away (but from running it I can see that the compiler is definitely not that clever and it behaves as expected - also the data is mostly to be looked at by human at this point and I think you'd notice if the whole thing suddenly took basically zero time). 

I might implement the increment logic you described, but to be fair - if the optimizer is clever and the actions in the loop are very trivial, it can guess that the counter will get incremented in each run  and only perform `counter += iteration` once - and the effect will be the same. So in a way you always have to be rely on the optimizer being "dumb enough" - and I observed in this case things are working fine.